packages/vitest/src/runtime/entry.ts

```diff
@@ -45,22 +52,31 @@ export async function run(files: string[], config: ResolvedConfig): Promise<void
     if (!files || !files.length)
       continue
 
-    await withEnv(environment, config.environmentOptions || {}, async () => {
-      for (const file of files) {
+    const filesByOptions = groupBy(files, ({ envOptions }) => JSON.stringify(envOptions))
+
+    for (const options of Object.keys(filesByOptions)) {
+      const files = filesByOptions[options]
+
+      if (!files || !files.length)
+        continue
+
+      await withEnv(environment, JSON.parse(options) || config.environmentOptions || {}, async () => {
```

Comment:
I think we can get parsed options from `files[0]`? To remove overhead of parsing.